File: src/BusinessDays.cs
using System;
using System.Collections.Generic;

namespace PaymentAlert
{
    /// <summary>주말과 공휴일을 제외한 영업일 계산.</summary>
    public class BusinessDayCalendar
    {
        readonly HashSet<DateTime> holidays;
        readonly HashSet<int> coveredYears;

        public BusinessDayCalendar(IEnumerable<DateTime> holidayDates, IEnumerable<int> years)
        {
            holidays = new HashSet<DateTime>();
            foreach (DateTime d in holidayDates) holidays.Add(d.Date);
            coveredYears = new HashSet<int>();
            foreach (int y in years) coveredYears.Add(y);
        }

        /// <summary>해당 연도의 공휴일 자료를 갖고 있는가.</summary>
        public bool HasYear(int year) { return coveredYears.Contains(year); }

        public bool IsHoliday(DateTime d) { return holidays.Contains(d.Date); }

        public bool IsBusinessDay(DateTime d)
        {
            if (d.DayOfWeek == DayOfWeek.Saturday || d.DayOfWeek == DayOfWeek.Sunday) return false;
            return !IsHoliday(d);
        }

        /// <summary>영업일이 아니면 다음 영업일로 민다. 영업일이면 그대로.</summary>
        public DateTime NextBusinessDayOrSame(DateTime d)
        {
            DateTime x = d.Date;
            int guard = 0;
            while (!IsBusinessDay(x))
            {
                x = x.AddDays(1);
                if (++guard > 400) throw new InvalidOperationException("영업일을 찾지 못했습니다. 공휴일 자료를 확인하세요.");
            }
            return x;
        }

        /// <summary>from에서 n영업일 앞선 날짜. n=3이면 3영업일 전.</summary>
        public DateTime SubtractBusinessDays(DateTime from, int n)
        {
            DateTime x = from.Date;
            int moved = 0;
            int guard = 0;
            while (moved < n)
            {
                x = x.AddDays(-1);
                if (IsBusinessDay(x)) moved++;
                if (++guard > 1000) throw new InvalidOperationException("영업일 역산에 실패했습니다.");
            }
            return x;
        }

        /// <summary>from(제외)부터 to(포함)까지의 영업일 수. to가 과거면 음수.</summary>
        public int BusinessDaysBetween(DateTime from, DateTime to)
        {
            DateTime a = from.Date, b = to.Date;
            if (a == b) return 0;

            int sign = b > a ? 1 : -1;
            DateTime lo = sign > 0 ? a : b;
            DateTime hi = sign > 0 ? b : a;

            int count = 0;
            DateTime x = lo.AddDays(1);
            while (x <= hi)
            {
                if (IsBusinessDay(x)) count++;
                x = x.AddDays(1);
            }
            return count * sign;
        }
    }
}

File: node-server/lib/businessdays.js
"use strict";
// src/BusinessDays.cs 를 그대로 옮긴 것. 결과가 달라지면 안 되므로 계산 순서까지 같게 둔다.
const D = require("./dates");

class BusinessDayCalendar {
  /**
   * @param {string[]} holidayDates 'YYYY-MM-DD' 목록
   * @param {number[]} years 공휴일 자료를 가진 연도
   */
  constructor(holidayDates, years) {
    this.holidays = new Set(holidayDates || []);
    this.coveredYears = new Set(years || []);
  }

  /** 해당 연도의 공휴일 자료를 갖고 있는가. */
  hasYear(year) { return this.coveredYears.has(year); }

  isHoliday(d) { return this.holidays.has(D.ymd(d)); }

  isBusinessDay(d) {
    const w = D.dow(d);
    if (w === 0 || w === 6) return false;
    return !this.isHoliday(d);
  }

  /** 영업일이 아니면 다음 영업일로 민다. 영업일이면 그대로. */
  nextBusinessDayOrSame(d) {
    let x = d;
    let guard = 0;
    while (!this.isBusinessDay(x)) {
      x = D.addDays(x, 1);
      if (++guard > 400) throw new Error("영업일을 찾지 못했습니다. 공휴일 자료를 확인하세요.");
    }
    return x;
  }

  /** from 에서 n 영업일 앞선 날짜. n=3 이면 3영업일 전. */
  subtractBusinessDays(from, n) {
    let x = from;
    let moved = 0;
    let guard = 0;
    while (moved < n) {
      x = D.addDays(x, -1);
      if (this.isBusinessDay(x)) moved++;
      if (++guard > 1000) throw new Error("영업일 역산에 실패했습니다.");
    }
    return x;
  }

  /** from(제외)부터 to(포함)까지의 영업일 수. to 가 과거면 음수. */
  businessDaysBetween(from, to) {
    const a = from.getTime(), b = to.getTime();
    if (a === b) return 0;

    const sign = b > a ? 1 : -1;
    const lo = sign > 0 ? from : to;
    const hi = sign > 0 ? to : from;

    let count = 0;
    let x = D.addDays(lo, 1);
    while (x.getTime() <= hi.getTime()) {
      if (this.isBusinessDay(x)) count++;
      x = D.addDays(x, 1);
    }
    return count * sign;
  }
}

module.exports = { BusinessDayCalendar };
